etebase: fix runtime crash due to wrong pydantic..

phaer opened the pull request:
...version. Upstream still uses pydantic 1, which it implicitly imports via fastapi. Overriding 
that and its pydantic-using dependencies fixes it, until upstream updates.

also:
* exposes the existing nixos integration test via passthru.nixosTests
* adds myself as a maintainer
* makes the package and pythonPackage used for the service configurable.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

phaer committed:
etebase: fix runtime crash due to wrong pydantic..

...version. Upstream still uses pydantic 1, which it
implicitly imports via fastapi. Overriding that and its
pydantic-using dependencies fixes it, until upstream updates.

phaer committed:
etebase: add phaer to maintainers

phaer committed:
modules/etebase-server: add package, pythonPackage

options, to keep the packaeges configurable

phaer commented:
@felschr Ideally, we could merge the package options soon, even if the rest of the changes require discussion. As it's currently broken and calendar sync can be quite critical, some users - like myself -  can't really upgrade since the last pyton-update.

With a package option one gains the capability to patch that single package without needing to vendor the whole module.

felschr reviewed (approved):
Looks all good to me. Thank you for the contribution!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1405

SuperSandro2000 commented on pkgs/servers/etebase/default.nix:
```diff
@@ -9,13 +8,22 @@
 
-buildPythonPackage rec {
+let
+  python = python3.override {
```
We should also use below python to make sure we are not mixing versions of Python or run into conflicts.

phaer replied:
WDYM? We are using this python for the default package. Users could still override for custom packages as needed?

SuperSandro2000 replied:
We are not using it for the other dependencies like django, etc.

SuperSandro2000 commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -46,6 +46,20 @@ in
+        description = lib.mdDoc "etebase-server package to use.";
+      };
+
+      pythonPackage = mkOption {
```
We should implement this by passing thru the interpreter used to build the package. Maybe that is already available at .python, not fully sure.

phaer replied:
Happy to do this, but wouldn't using "the interpreter used to build the package" imply infinite recursion?

The default for package is cfg.pythonPackages.pkgs.etebase-server, so I can't really evaluate that to access `.pythonModule` before I know the python interpreter to use?

SuperSandro2000 replied:
You should be able to access cfg.package.python without triggering an infinite recursion as this is just passed thru, too.

SuperSandro2000 commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -5,8 +5,8 @@ with lib;
-  pythonEnv = pkgs.python3.withPackages (ps: with ps;
-    [ etebase-server daphne ]);
+  pythonEnv = cfg.pythonPackage.withPackages (ps: with ps;
+    [ cfg.package daphne ]);
```
I have my doubts that this is creating a proper env. We should add this in the package instead.

phaer replied:
This has been here before this PR, which fixes a non-functional service. Happy to change this in a follow up, it should also use uwsgi as upstream.

It creates a proper venv insofar as it works. (see e.g. nixos tests)

phaer commented:
@SuperSandro2000 I think i addressed all of your wishes in the recent commits and made it a proper python package while doing so.

Would be nice to have this merged soonish as this service has been broken on unstable for well over a month now.

phaer committed:
etebase: use dependencies from overriden python...

interpreter, as per review comment.

phaer committed:
etebase: make proper package...

and remove the ad-hoc python environment.
Also remove daphne and use uvicorn just as upstream does

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1491

SuperSandro2000 commented on pkgs/servers/etebase/default.nix:
```diff
@@ -29,32 +27,46 @@ buildPythonPackage rec {
     chmod +x $out/bin/etebase-server
   '';
 
   passthru.updateScript = nix-update-script {};
+  passthru.python = python;
+  # PYTHONPATH of all dependencies used by the package
+  passthru.pythonPath = python.pkgs.makePythonPath propagatedBuildInputs;
+  passthru.tests = {
+    nixosTest = nixosTests.etebase-server;
+  };
```
```suggestion
  passthru = {
    updateScript = nix-update-script {};
    python = python;
    # PYTHONPATH of all dependencies used by the package
    pythonPath = python.pkgs.makePythonPath propagatedBuildInputs;
    tests.nixosTest = nixosTests.etebase-server;
```
